django/db/models/functions/date.py

```diff
@@ -0,0 +1,214 @@
+    lookup_name = 'second'
+
+
+DateField.register_lookup(ExtractYear)
+DateField.register_lookup(ExtractMonth)
+DateField.register_lookup(ExtractDay)
+DateField.register_lookup(ExtractWeekDay)
+
+TimeField.register_lookup(ExtractHour)
+TimeField.register_lookup(ExtractMinute)
+TimeField.register_lookup(ExtractSecond)
+
+DateTimeField.register_lookup(ExtractYear)
+DateTimeField.register_lookup(ExtractMonth)
+DateTimeField.register_lookup(ExtractDay)
+DateTimeField.register_lookup(ExtractWeekDay)
+DateTimeField.register_lookup(ExtractHour)
+DateTimeField.register_lookup(ExtractMinute)
+DateTimeField.register_lookup(ExtractSecond)
+
+ExtractYear.register_lookup(YearExact)
+ExtractYear.register_lookup(YearGt)
+ExtractYear.register_lookup(YearGte)
+ExtractYear.register_lookup(YearLt)
+ExtractYear.register_lookup(YearLte)
+
+
+class TruncBase(Transform):
+    arity = 1
+    kind = None
+    tzinfo = None
+
+    def as_sql(self, compiler, connection):
+        inner_sql, inner_params = compiler.compile(self.lhs)
+        # Escape any params because trunc_sql will format our string
+        inner_sql = inner_sql.replace('%s', '%%s')
+        if isinstance(self.output_field, DateTimeField):
+            tzname = None
+            if settings.USE_TZ:
+                if self.tzinfo is None:
+                    tzname = timezone.get_current_timezone_name()
+                else:
+                    tzname = timezone._get_timezone_name(self.tzinfo)
+            sql, params = connection.ops.datetime_trunc_sql(self.kind, inner_sql, tzname)
+        elif isinstance(self.output_field, DateField):
+            sql = connection.ops.date_trunc_sql(self.kind, inner_sql)
+            params = []
+        else:
+            raise ValueError('Trunc only valid on Date or DateTime fields')
+        return sql, inner_params + params
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        copy = super(TruncBase, self).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        field = copy.lhs.output_field
+        # DateTimeField is a subclass of DateField so this works for both.
+        assert isinstance(field, DateField), (
+            "%r isn't a DateField or DateTimeField." % field.name
+        )
+        return copy
+
+    def convert_value(self, value, expression, connection, context):
+        if isinstance(self.output_field, DateTimeField):
+            if settings.USE_TZ:
+                if value is None:
+                    raise ValueError(
+                        "Database returned an invalid datetime value. "
+                        "Are time zone definitions for your database and pytz installed?"
+                    )
+                value = value.replace(tzinfo=None)
+                value = timezone.make_aware(value, self.tzinfo)
+        elif isinstance(self.output_field, DateField) and isinstance(value, datetime):
+            value = value.date()
```

Comment:
These two lines make me sad... I had managed to delineate dates and datetimes cleanly by separating QuerySet.dates() and QuerySet.datetimes(). This new code makes it possible to accidentially treat datetimes and dates, which is usually signals incorrect database modeling and bugs waiting to happen (especially for Europeans who can ignore the different between <local time>.date() and <UTC>.date() as long as nothing happens within two hours of midnight, which is often the case for business appps).